[2단계 - 자동차 경주] 리브(김민주) 미션 제출합니다

Minjoo522 opened the pull request:
안녕하세요, 찰리 😌
이번 단계에서는 1단계에서 추가로 남겨주신 댓글과 코멘트를 반영하고 객체의 역할과 책임에 맞게 추가로 새로운 도메인을 만들었습니다.

### `RacingCarController` 이름
`RacingCarController` 이름은 말씀 주신 바와 같이 MVC 패턴을 사용하며 습관적으로 사용한 이름이었습니다.
개발자의 입장이 아니라 현실이라면 `RacingCarController`를 실행한다고 할 때 어떤 이름을 쓸지 더 고민해 보았습니다.
컨트롤러라고 하지 않는 것은 아니지만 보통 `게임기를 실행한다.`라고 한다고 생각합니다.
그래서 `RacingCarGameMachine`이라고 이름을 변경해 보았습니다. 😁

### `Map<String, Integer>` 사용을 주저했던 이유
메서드 사용성 측면에서 생각 했을 때, `Map<String, Integer>`를 리턴한다면 key와 value에 어떤 값이 들어 있는지 직관적으로 파악하기 어렵다고 생각했습니다.
만약, `Map<CarName, Position>`이었다면 어떤 값이 들어 있는지 데이터 타입을 보고 바로 알 수는 있지만, View에서 사용하기 때문에 View가 domain에 의존하게 되어 이 또한 좋지 않다고 판단하였습니다.
그리고 Map에 대해서 다시 확인해 보았는데 `순서를 보장하는 Map은 LinkedHashMap`이었다는 것을 깨달았습니다. 😅

### 테스트 코드 예외 메세지 `String.format` 사용
1단계에서 댓글로 남겨주신 것에서 예외 메세지에 String.format을 사용하는 것이 더 적합하다고 생각하여 프로덕션 코드에 반영했습니다. 😆

```java
String.format("자동차 대수는 %d대 이상 %d대 이하여야 합니다.", MIN_CARS_SIZE, MAX_CARS_SIZE)
```

다만, 현업에서 테스트 코드에도 동일하게 String.format을 사용하는지 궁금합니다. 🤔

저는  테스트 코드는 현 시점에서의 프로그램 규칙을  반영해야 한다고 생각합니다.
실제 프로그램의 규칙은 변경되지 않았는데 누군가의 실수로 상수가 변경되었을 때, 테스트 코드는 실패해야 한다고 생각합니다.
그래서 테스트 코드에는 String.format을 사용하지 않았습니다.
찰리는 어떻게 생각하시는지 의견이 궁금합니다. 😊

### `TryCount`와 `CarName` 도메인 생성
기존에는 시도 횟수가 최소값 ~ 최대값 사이인지 유효성 검사를 `InputView`에서 시행했고 int 타입 그대로를 사용했습니다.
해당 유효성 검사가 View에서 할 책임은 아니라고 생각하여 `TryCount`를 분리하여 주었습니다.
그리고 `TryCount`에 끝났는지 여부를 질문하여 한 라운드를 실행하도록 변경했습니다.

```java
    private void proceedRounds(final Cars cars, final TryCount tryCount) {
        OutputView.printResultSubject();
        while (tryCount.isNotFinish()) {
            tryCount.tryOne();
            proceedOneRound(cars);
        }
    }
```

`CarName` 역시 `Car`에서 자동차 이름의 유효성 검사까지 하는 것은 과도한 책임이라는 판단 하에 새로운 도메인을 추가해 주었습니다.

### result 모아서 `OutputView`에 전달
현 상황에서 result를 모아서 `OutputView`에 전달하면, `OutputView`에서 전달 받아야 할 데이터 형식이 List<List<TryCount>>가 됩니다.
두 번이나 List로 묶여 있는 데이터를 `OutputView`에서 한 번에 풀어서 사용하면 `OutputView`의 역할이 너무 복잡해지는 것은 아닐까 고민됩니다.
혹시 찰리는 이 부분에 대해서 어떻게 생각하시는지 궁금합니다. 🤔

Minjoo522 committed:
fix: 예외 메세지 오류 수정

Minjoo522 committed:
fix: 예외 메세지 오류 수정

Minjoo522 committed:
rerfactor: 예외 메세지의 숫자 상수 사용할 수 있도록 String.format 사용

Minjoo522 committed:
rerfactor: 테스트코드 DisplayName 매끄럽게 읽히도록 개선

Minjoo522 committed:
rerfactor: getCars, getTryCount 메서드 의미있는 이름으로 개선

Minjoo522 committed:
rerfactor: 메서드 순서 위에서 아래로 읽히도록 개선

Minjoo522 committed:
feat: tryCount 원시 값을 포장할 수 있는 객체 생성

Minjoo522 committed:
refactor: tryCount를 원시 값(int) 그대로 사용하지 않고 TryCount 객체에 질의하여 라운드들을 진행 할 수 있도록 개선

Minjoo522 committed:
rename: TryCount 객체로 필요한 유효성 검사를 옮기며 InputView에서 필요 없어진 테스트 제거

Minjoo522 committed:
refactor: TryCount 생성 시 0이나 음수가 들어온 케이스 각각 테스트 할 수 있도록 개선

Minjoo522 committed:
feat: 자동차 이름을 포장할 수 있는 CarName 객체 추가

Minjoo522 committed:
refactor: CarName class를 record로 변경

Minjoo522 committed:
fix: CarNameTest 자동차 이름 길이 검사 테스트에서 CarName을 생성하도록 수정

Minjoo522 committed:
refactor: 자동차 이름을 CarName 도메인을 통해 생성하고 Car에서도 CarName을 사용하도록 개선

Minjoo522 committed:
rename: RacingCarController의 이름을 더 의미 있는 이름으로 변경

Minjoo522 committed:
rename: tryCount가 0이 아닌지 확인하는 메서드의 이름을 더 적합한 이름으로 변경

Gomding commented:
> ### RacingCarController 이름
> RacingCarController 이름은 말씀 주신 바와 같이 MVC 패턴을 사용하며 습관적으로 사용한 이름이었습니다.
개발자의 입장이 아니라 현실이라면 RacingCarController를 실행한다고 할 때 어떤 이름을 쓸지 더 고민해 보았습니다.
컨트롤러라고 하지 않는 것은 아니지만 보통 게임기를 실행한다.라고 한다고 생각합니다.
그래서 RacingCarGameMachine이라고 이름을 변경해 보았습니다. 😁

좋은 의견이네요! 💯 
Controller 라고 이름을 지었을 때 개발자라면 공통적으로 떠올리는 것이 있다는 장점이 있지만
지금 미션에서는 굳이 그렇게 할 필요는 없다고 생각했어요 :)

Gomding commented:
> ### Map<String, Integer> 사용을 주저했던 이유
> 메서드 사용성 측면에서 생각 했을 때, Map<String, Integer>를 리턴한다면 key와 value에 어떤 값이 들어 있는지 직관적으로 파악하기 어렵다고 생각했습니다.
> 만약, Map<CarName, Position>이었다면 어떤 값이 들어 있는지 데이터 타입을 보고 바로 알 수는 있지만, View에서 사용하기 때문에 View가 domain에 의존하게 되어 이 또한 좋지 않다고 판단하였습니다.
그리고 Map에 대해서 다시 확인해 보았는데 순서를 보장하는 Map은 LinkedHashMap이었다는 것을 깨달았습니다. 😅

크.. 좋은 관점이네요  👍 👍 👍

Gomding commented:
> ### 테스트 코드 예외 메세지 String.format 사용
> 1단계에서 댓글로 남겨주신 것에서 예외 메세지에 String.format을 사용하는 것이 더 적합하다고 생각하여 프로덕션 코드에 반영했습니다. 😆
> 
> String.format("자동차 대수는 %d대 이상 %d대 이하여야 합니다.", MIN_CARS_SIZE, MAX_CARS_SIZE)
> 다만, 현업에서 테스트 코드에도 동일하게 String.format을 사용하는지 궁금합니다. 🤔
> 
> 저는 테스트 코드는 현 시점에서의 프로그램 규칙을 반영해야 한다고 생각합니다.
> 실제 프로그램의 규칙은 변경되지 않았는데 누군가의 실수로 상수가 변경되었을 때, 테스트 코드는 실패해야 한다고 생각합니다.
> 그래서 테스트 코드에는 String.format을 사용하지 않았습니다.
> 찰리는 어떻게 생각하시는지 의견이 궁금합니다. 😊

String.format을 사용하는 곳은 예외 메시지에 대한 검증인데
테스트에서 예외에 적절한 메시지인지 검증할 때 2나 10이라는 숫자가 중요할지 고민해보면 어떨까요? 😄 
매번 예외 메시지 검증부를 수정해야하는것과 트레이드 오프도 함께 생각해보아요!

Gomding commented:
> ### TryCount와 CarName 도메인 생성
> 기존에는 시도 횟수가 최소값 ~ 최대값 사이인지 유효성 검사를 InputView에서 시행했고 int 타입 그대로를 사용했습니다.
> 해당 유효성 검사가 View에서 할 책임은 아니라고 생각하여 TryCount를 분리하여 주었습니다.
> 그리고 TryCount에 끝났는지 여부를 질문하여 한 라운드를 실행하도록 변경했습니다.
```java
    private void proceedRounds(final Cars cars, final TryCount tryCount) {
        OutputView.printResultSubject();
        while (tryCount.isNotFinish()) {
            tryCount.tryOne();
            proceedOneRound(cars);
        }
    }
```
> CarName 역시 Car에서 자동차 이름의 유효성 검사까지 하는 것은 과도한 책임이라는 판단 하에 새로운 도메인을 추가해 주었습니다.

원시값 포장을 사용해주셨군요 👍 
이왕 포장한김에 Car에 있는 Position도 포장해보죠! 😄

Gomding commented:
> ### result 모아서 OutputView에 전달
> 현 상황에서 result를 모아서 OutputView에 전달하면, OutputView에서 전달 받아야 할 데이터 형식이 List<List>가 됩니다.
> 두 번이나 List로 묶여 있는 데이터를 OutputView에서 한 번에 풀어서 사용하면 OutputView의 역할이 너무 복잡해지는 것은 아닐까 고민됩니다.
> 혹시 찰리는 이 부분에 대해서 어떻게 생각하시는지 궁금합니다. 🤔

분명 이중 리스트는 조금 복잡하지만 사용이 꺼려질만큼 복잡할까요? 😄 

매번 OutputView를 호출하지만 간결한 값
VS
한 번만 OutputView를 호출하지만 약간 복잡한 값

트레이드 오프를 고민해보면 좋겠어요!

Gomding commented on src/main/java/racingcar/domain/CarName.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.regex.Pattern;
+
+public record CarName(String name) {
+    private static final Pattern NAME_PATTERN = Pattern.compile("[^ㄱ-ㅎ가-힣a-zA-Z0-9_-]");
```
저번에 말씀을 안드렸는데
한글도 가능하게 해달라는 요청을 들어주셨군요 👍 👍 👍

Gomding commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,9 +1,9 @@
 public class Application {
     public static void main(String[] args) {
-        new RacingCarController().run();
+        new RacingCarGameMachine().run();
```
2단계 미션 요구사항에 이런 부분이 있던데

리브도 한 번 구조 리팩토링을 해보면 어떨까요!

![image](https://github.com/woowacourse/java-racingcar/assets/57378410/c34f75c7-0ca7-4f00-b670-6659a887d4b0)

Minjoo522 replied:
관련 요구 사항 예시와 동일하게 하지 않았던 이유에 대해서 설명을 덧붙였어야 했는데 부족했습니다. 😭

이유는, 1단계에서 자체적으로 추가 했던 요구 사항 중에 `사용자의 입력에 대한 예외가 발생하면 해당 입력을 다시 받을 수 있도록 돌아간다.`로 인해서 완전히 동일하게 구현하지 않는게 좋겠다는 판단을 내렸습니다.

만약 현재 요구 사항을 유지하며 예시와 동일하게 구현한다면 자동차 이름을 입력 받고 `InputView`에서 검증하는 유효성 검사에서 예외가 발생했을 경우에만 입력을 다시 받을 수 있고 `CarName`이나 `Cars` 객체에서 검증하는 검사는 `tryCount`까지 입력 받은 후 확인하게 됩니다.
또한,  예외 발생시 다시 돌아가기 위해서 `InputView`의 `getCarNames`, `getTryCount`에서 예외가 발생 했는지 try-catch 하고 반복해 주어야 합니다.
그렇다면 `예외가 발생 했을 때 다시 입력 받도록 되돌아가게 하는 것이 InputView의 역할인가?` 라는 또 다른 고민이 생겼고, 저는 controller의 역할이라고 생각했습니다.

제 코드에서는 main에서 view를 사용하지 않고 view와 domain의 데이터 흐름을 제어하는 controller가 있어야만 한다는 결론을 내게 되었습니다.

찰리가 남겨주신 코멘트 덕분에 다시금 고민해보며, 입력 받는 방식을 정하는 `ConsoleReader`는 여러 번 사용되기 때문에 controller 생성시 주입받고 controller의 final 필드에 저장하여 사용하는 것이 좋겠다고 여겨 반영해 보았습니다. 😊

Gomding replied:
리브의 의도가 담긴 부분이었군요 💯 💯 💯 💯 
설명해주셔서 감사해요 😄 
충분히 이해가 되는 부분이라 수정하지 않으셔도 됩니다!

Gomding commented on src/main/java/racingcar/controller/RacingCarGameMachine.java:
```diff
@@ -1,50 +1,54 @@
+    private Cars initCars() {
         try {
             final List<Car> cars = createCars(InputView.readCarNames(new ConsoleReader()));
             return new Cars(cars);
```
Cars 를 생성할 때 인자로 자동차 이름들을 바로 넘기는건 어떻게 생각하시나요? 😄 

```java
Cars.of(carNames);
```

Minjoo522 replied:
좋은 의견 감사합니다!
자동차 이름들을 바로 넘기니 비즈니스 로직이 도메인으로 이동할 수 있게 되어 더 좋은 것 같습니다.
또, 생성자를 직접 사용하는 것 보다 의도를 명확히 표현할 수 있게 된 것 같아요!
바로 반영 하였습니다! 😊

Gomding replied:
외부에 더이상 Car를 노출하지 않는다는 장점도 있어요! 😄 
실제로 RacingCarGameMachine 의 import 를 보면 Car에 대한 의존이 사라진것을 볼 수 있어요

(import 구간을 확인해보면 model, controller, view 간에 의존이 제대로 이뤄지고 있는지 체크해볼 수 있어요!)

```java
package racingcar.controller;

import java.util.ArrayList;
import java.util.List;
import racingcar.domain.Cars;
import racingcar.domain.TryCount;
import racingcar.dto.CarStatus;
import racingcar.util.RandomNumberGenerator;
import racingcar.view.InputView;
import racingcar.view.OutputView;
import racingcar.util.ConsoleReader;

public class RacingCarGameMachine {
```

Gomding commented on src/main/java/racingcar/controller/RacingCarGameMachine.java:
```diff
@@ -1,50 +1,54 @@
         OutputView.printResultSubject();
-        IntStream.rangeClosed(1, tryCount)
-                .forEach(i -> proceedOneRound(cars));
+        while (tryCount.isNotFinish()) {
```
아래와 같이 tryCount에게 요청해보는 방법은 어떠신가요! 😄 
(장단점을 비교해보시면 좋겠어요!)

```java
tryCount.execute(() -> proceedOneRound(cars));
```

Minjoo522 replied:
좋은 방법을 알려주셔서 감사합니다! ☺️
이렇게 하니, 한 라운드를 돌 때마다 isNotFinish를 호출하지 않고도 반복할 수 있고 tryCount의 필드도 final로 유지할 수 있게 되었습니다. 👍
알려주신 방법의 단점은 `tryCount 숫자만큼 반복해서 proceedOneRound를 실행한다는 것을 TryCount의 execute를 봐야만 알 수 있다`라고 생각하는데, controller에서 `proceedRounds`, `proceedOneRound`로 나뉘어 있어 추측이 가능할 것으로 보여서 장점을 상쇄할만한 단점은 아닌 것 같습니다.

Gomding commented on src/main/java/racingcar/controller/RacingCarGameMachine.java:
```diff
@@ -1,50 +1,54 @@
     private List<Car> createCars(final List<String> carNames) {
         return carNames.stream()
-                .map(Car::new)
+                .map(carName -> new Car(new CarName(carName)))
```
Car를 생성할 때도 carName을 생성자에 바로 전달해보면 어떨까요? 😄

Minjoo522 replied:
바로 반영했습니다! 😆

Gomding commented on src/main/java/racingcar/domain/TryCount.java:
```diff
@@ -0,0 +1,32 @@
+
+public class TryCount {
+    private static final int MIN_TRY_COUNT = 1;
+    private static final int MAX_TRY_COUNT = 1_000;
```
언더바로 숫자의 가독성 높여주기 💯 💯 💯 💯 💯

Gomding commented on src/test/java/racingcar/domain/CarNameTest.java:
```diff
@@ -0,0 +1,38 @@
+    @ParameterizedTest
+    @NullAndEmptySource
+    @ValueSource(strings = {" ", "  ", "\t", "\n"})
+    @DisplayName("[Exception] 자동차 이름이 공백이나 null이면 예외를 던진다")
```
꼼꼼한 테스트 👍

Gomding commented on src/test/java/racingcar/domain/TryCountTest.java:
```diff
@@ -0,0 +1,28 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class TryCountTest {
```
validation이외에 public 메서드는 간단한 로직이지만
테스트 작성해보면 좋겠어요! 😄

Minjoo522 replied:
위에서 아래와 같이 코드를 변경하며,
기존에 있던 validate 이외의 메소드는 모두 사라지게 되었습니다.
```java
tryCount.execute(() -> proceedOneRound(cars));
```

현재 execute 메소드는 아래와 같은데

```java
public void execute(Runnable action) {
    IntStream.rangeClosed(1, tryCount)
            .forEach(i -> action.run());
}
```

잘 만들어져 있는 스트림 라이브러리를 사용해서 실행만 시켜주는 코드이기 때문에 테스트 하지 않아도 괜찮다고 생각했습니다.
찰리의 의견은 어떠신가요? 🤔

Gomding replied:
내부적으로는 `tryCount의 횟수만큼 반복한다` 라는 내용을 담고있으니
이것에 대한 테스트를 작성해보면 어떨까요..!

Minjoo522 replied:
잘 짜인 테스트 코드는 문서화의 역할도 할 수 있다는 사실을 간과했네요!
바로 추가하였습니다. 😆

Gomding commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -47,16 +49,16 @@ public List<String> getWinners() {
-                .max(Car::compareTo)
-                .orElseThrow(IllegalArgumentException::new);
-    }
 }
```
Cars의 move() 메서드는 랜덤값에 의존하고있어 테스트할 수 없는 상황으로 보이네요!

아래글을 학습하고 테스트 작성해보면 좋겠어요~
- https://tecoble.techcourse.co.kr/post/2020-05-17-appropriate_method_for_test_by_interface/

Minjoo522 replied:
감사합니다!
`NumberGenerator` 인터페이스를 생성하고 `RandomNumberGenerator`가 구현하도록 만들었습니다.
그리고 `Cars`의 `move` 메서드에서 `NumberGenerator`를 매개변수로 받도록 변경해 보았습니다.

Gomding reviewed (changes requested):
안녕하세요 리브!

2단계 빠르게 제출해주셨군요 😮 
꼼꼼하게 답변까지 달아주셔서 감사합니다!!

몇가지 코멘트 남겼으니 의견주시길 바래요 😄 
궁금한 점 있으면 코멘트나 DM 남겨주세요 🙏

Minjoo522 committed:
refactor: ConsoleReader를 주입 받아 controller를 생성하도록 변경

Minjoo522 committed:
refactor: TryCount에 요청하여 라운드를 요청된 시도 횟수만큼 반복하도록 개선

Minjoo522 committed:
refactor: 자동차 이름을 직접 받아서 Car를 생성하도록 개선

Minjoo522 committed:
refactor: 자동차 이름 리스트를 직접 받아서 Cars를 생성하도록 개선

Minjoo522 committed:
refactor: 자동차들을 이동시킬 숫자를 생성할 NumberGenerator를 주입 받을 수 있도록 개선

Minjoo522 committed:
fix: 지워진 맨 마지막 줄 추가

Minjoo522 committed:
refactor: position 원시 값을 포장하는 Position 객체를 사용하도록 개선

Minjoo522 committed:
refactor: 테스트 코드의 예외 메세지 안의 숫자들 상수를 이용하도록 변경

Minjoo522 committed:
refactor: 각 라운드의 결과를 모아 두었다가 한번에 OutputView에 전달하도록 개선

Minjoo522 committed:
test: Cars move메소드 테스트 코드 추가

Minjoo522 committed:
comment: 가독성을 위한 주석 추가

Minjoo522 committed:
rename: 혼동을 줄 수 있는 변수명 변경

Minjoo522 commented:
안녕하세요, 찰리! 😆
리뷰 남겨 주셔서 감사합니다. ☺️

테스트 예외 메세지와 관련하여 코멘트를 보고 다시 고민해보니 숫자가 중요할 것 같진 않다라고 생각이 바뀌었습니다.
테스트 코드의 역할을 다시 고민해 보니 메소드가 잘 작동하는지 여부가 예외 메세지의 숫자보다는 더 중요하다는 생각이 들었습니다.
저번에 제가 예외 메세지를 각 객체로 옮기면서 숫자가 변경되는 실수를 했던 것처럼 오히려 상수 값에 대한 요구 사항이 변경되었을 때 예외 메세지를 하나하나 수정하는 것이 더 번거로울 것 같다는 생각이 들었습니다.
지금은 테스트 코드가 몇개 없지만 만약 더 많아진다면, 수정에 드는 비용이 커질 것 같습니다. 😅

> 매번 OutputView를 호출하지만 간결한 값
> VS
> 한 번만 OutputView를 호출하지만 약간 복잡한 값

위에 대해서도 정말 제가 생각하는 것처럼 후자가 복잡할지 직접 보고 싶어서 변경해 보았는데,
실제로 변경하고 보니 `proceedOneRound` 메소드가 이름에 맞는 역할만을 수행하게 되었다고 생각했습니다.
이전에는 한 라운드를 진행하고 print까지 해주었었네요. 😅

`Car`의 position도 포장해 주었는데,
이렇게 만들어주니 메소드 내부 구현을 읽을 때 숫자가 경계점 이상이면 포지션을 증가시킨다는 것을 쉽게 읽을 수 있어서 좋았습니다.
(사실 `position++`도 매우 짧아서 쉽게 해석할 수 있는 코드이긴 하지만요 😅)
```java
public void move(final int number) {
    if (number >= CAR_MOVE_THRESHOLD) {
        position = position.increasePosition();
    }
}
```

그동안 제가 트레이드 오프, 장단점을 신중히 고려해보지 않았다는 것을 깨달을 수 있었던 리뷰였습니다. 😆

Gomding commented:
> 테스트 예외 메세지와 관련하여 코멘트를 보고 다시 고민해보니 숫자가 중요할 것 같진 않다라고 생각이 바뀌었습니다.
> 테스트 코드의 역할을 다시 고민해 보니 메소드가 잘 작동하는지 여부가 예외 메세지의 숫자보다는 더 중요하다는 생각이 들었습니다.
> 저번에 제가 예외 메세지를 각 객체로 옮기면서 숫자가 변경되는 실수를 했던 것처럼 오히려 상수 값에 대한 요구 사항이 변경되었을 때 예외 메세지를 하나하나 수정하는 것이 더 번거로울 것 같다는 생각이 들었습니다.
> 지금은 테스트 코드가 몇개 없지만 만약 더 많아진다면, 수정에 드는 비용이 커질 것 같습니다. 😅

맞습니다 😄 
하지만 이전에 리브의 말대로 `상수의 변경에 대해 테스트도 정확하게 검증되고 있는가` 이것도 중요한 사항이에요!
그래서 적절하게 적용하는것이 중요하다고 생각해요 😃

Gomding commented on src/main/java/racingcar/controller/RacingCarGameMachine.java:
```diff
@@ -0,0 +1,62 @@
+
+    private void proceedOneRound(final Cars cars) {
+        cars.move(new RandomNumberGenerator());
+        results.add(cars.result());
```
매 라운드의 결과를 기록해놓기 👍

Gomding commented on src/main/java/racingcar/controller/RacingCarGameMachine.java:
```diff
@@ -0,0 +1,62 @@
+
+public class RacingCarGameMachine {
+    private final ConsoleReader reader;
+    private final List<List<CarStatus>> results = new ArrayList<>();
```
`List<RoundResult> results = new ArrayList<>();` 
이렇게 바꿔보면 이중 리스트의 복잡함을 조금 해소할 수 있어보여요 😄 
List<CarStatus> 하나가 어떤것을 의미하는지도 명확하구요!

물론 클래스 구조가 복잡해진다는 트레이드 오프가 존재합니다!

리브는 어떻게 생각하시나요?

Minjoo522 replied:
`RoundResult`를 `dto`로 생성하여 `Cars`의 `result()` 메서드에서 리턴하도록 개선해 보았습니다.
이렇게 만들고 보니 `result()` 메소드가 `List<CarStatus>`를 리턴할 때보다 더 명확한 역할을 하는 것처럼 보이네요.

```java
public RoundResult result() {
    return cars.stream()
            .map(Car::getCarStatus)
            .collect(Collectors.collectingAndThen(Collectors.toList(), RoundResult::new));
}
```

사용할 때 `value()`를 다시 꺼내서 처리해 주어야 한다는 불편함은 조금 있지만,
가독성 측면에서 얻는 이점이 훨씬 크다고 생각합니다. 😊

```java
private static void printResult(final RoundResult result) {
    result.value().stream()
            .map(OutputView::makeCarStatusResult)
            .forEach(System.out::println);
    System.out.println();
}
```

Gomding commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,69 +1,43 @@
 
     public CarStatus getCarStatus() {
-        return new CarStatus(name, position);
+        return new CarStatus(name.name(), position.position());
```
원시값을 포장한 CarName, Position 의 내부에 있는 것은 `값` 이니 
필드 이름도 `value` 로 지어보면 덜 어색하지 않을까요! 😄 

`name.name()` 에서 어색함이 느껴져서요!

```suggestion
        return new CarStatus(name.value(), position.value());
```

Minjoo522 replied:
좋은 의견 감사합니다!
바로 반영했습니다. 😆

Gomding commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -28,12 +36,16 @@ private void validateNotDuplicate(final List<Car> cars) {
 
-    public void move() {
-        cars.forEach(car -> car.move(RandomNumberGenerator.generate(MAX_RANDOM_NUMBER)));
+    public void move(NumberGenerator generator) {
```
외부에서 NumberGenerator 인터페이스 타입을 받아
랜덤에 대한 의존 제거 👍 👍 👍 👍

Gomding commented on src/main/java/racingcar/util/NumberGenerator.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.List;
+
+public interface NumberGenerator {
```
사소한 의견이라 반영하지 않으셔도 됩니다~

NumberGenerator를 자동차 게임에 대입하여 이름을 지어보는건 어떠신가요! 😄  
ex) GasStation, RefuelGenerator

Minjoo522 replied:
의견 주신 대로 변경하고 보니 훨씬 의미 있는 코드가 되었습니다.
감사합니다 😆

Gomding commented on src/test/java/racingcar/util/TestNumberGenerator.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.List;
+
+public class TestNumberGenerator implements NumberGenerator {
```
테스트 패키지에 생성해주셨군요 👍 👍 👍

Gomding commented on src/main/java/racingcar/domain/Position.java:
```diff
@@ -0,0 +1,8 @@
+public record Position(int position) {
+
+    public Position increasePosition() {
+        return new Position(position + 1);
```
불변을 위해 이렇게 구현해주셨을까요? 👍

Minjoo522 replied:
네!
원시값 포장을 하는 큰 이유 중 하나가 불변성 유지라고 생각했습니다. 😊

Gomding commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,69 +1,43 @@
-        this.name = name.trim();
+    private Car(final CarName name) {
+        this.name = name;
+        this.position = new Position(0);
```
0을 가진 Position 객체를 여러번 사용될 것으로 예상돼요!
`Position.ZERO` 이런식으로 객체를 재활용 해보면 어떨까요!

```suggestion
        this.position = Position.ZERO;
```

Minjoo522 replied:
좋은 의견 감사합니다!
바로 반영해 보았는데 의도대로 잘 반영된 것인지 확인하여 주시면 감사하겠습니다. 😊

Gomding replied:
리브가 만족하면 된다고 생각합니다 😄 
(제가 생각했던 방향은 맞아요 ㅎㅎ!)

Gomding reviewed (changes requested):
안녕하세요 리브!
리뷰 깔끔하게 반영해주셨네요!
의견까지 말씀해주셔서 감사드려요 🙇 

추가로 몇 가지 코멘트 남겼으니 확인부탁드려요 🙏 
궁금한 점 있으면 언제든 DM 남겨주세요!

Minjoo522 committed:
rename: CarName, Position의 값 이름 value로 변경
- 호출 시 name.name()이 아닌 name.value()로 호출하게 되어 가독성 향상

Minjoo522 committed:
test: TryCount의 execute 메소드 테스트 추가

Minjoo522 committed:
refactor: RoundResult DTO 생성하여 Cars의 result에서 RoundResult를 리턴하도록 개선
- 컨트롤러에서 List<RoundResult>로 관리하며 가독성 개선

Minjoo522 committed:
refactor: new Position(0)을 필드에 상수로 선언

Minjoo522 committed:
refactor: NumberGenerator의 이름 RefuelGenerator로 변경
- 인터페이스 구현체들도 네이밍 유사하게 변경
- Car의 move에서도 number가 아닌 fuel을 변수명으로 사용하도록 변경하여 통일성을줌

Minjoo522 commented:
안녕하세요, 찰리! 😆
빠르게 리뷰해 주셔서 정말 감사합니다.👍
리뷰를 통해서 코드가 계속 발전해 나가는 것을 보니 정말 즐겁고 또 즐거워서 잠이 오질 않네요 😉💤
이제 월요일인데, 이번 한 주도 즐거운 한 주가 되시길 바랍니다! 😆

Gomding commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -1,48 +1,48 @@
     }
 
     @ParameterizedTest
     @MethodSource("InputCarNames")
```
MethodSource 활용 👍 
이미 Junit을 활용한 테스트에 익숙하신 것같군요 💯

Gomding reviewed (approved):
안녕하세요 리브!
많은 고민과 학습을 통해 잘 구현해주셨네요~
이번 미션은 여기서 종료하겠습니다 :)

추가로 궁금하거나 고민있으시면 DM 남겨주세요 😄 

다음 미션도 화이팅입니다 🔥
